Sources/MlemMiddleware/FeedLoaders/Generics/LoadingActor.swift

```diff
@@ -0,0 +1,166 @@
+//
+//  LoadingActor.swift
+//  MlemMiddleware
+//
+//  Created by Eric Andrews on 2024-10-27.
+//
+
+import Foundation
+
+enum LoadingError: Error {
+    case noTask
+}
+
+enum LoadingResponse<Item: FeedLoadable> {
+    /// Indicates a successful load with more items available to fetch
+    case success([Item])
+    
+    /// Indicates a successful load with no more items available to fetch
+    case done([Item])
+    
+    /// Indicates the load was ignored due to an existing ongoing load
+    case ignored
+    
+    /// Indicates the load was cancelled
+    case cancelled
+    
+    var description: String {
+        switch self {
+        case let .success(items): "success (\(items.count))"
+        case let .done(items): "done (\(items.count))"
+        case .ignored: "ignored"
+        case .cancelled: "cancelled"
+        }
+    }
+}
+
+public class Fetcher<Item: FeedLoadable> {
+    var page: Int
+    private var cursor: String?
+    
+    init (page: Int = 0) {
+        self.page = page
+    }
+    
+    /// Helper struct bundling the response from a fetchPage or fetchCursor call
+    struct FetchResponse<Item: FeedLoadable> {
+        /// Items returned
+        public let items: [Item]
+        
+        /// Cursor used to fetch this response, if applicable
+        public let prevCursor: String?
+        
+        /// New cursor, if applicable
+        public let nextCursor: String?
+    }
+    
+    /// Fetches the next page of items
+    func fetch() async throws -> LoadingResponse<Item> {
+        do {
+            if let cursor, page > 0 {
+                print("[\(Item.self) Fetcher] loading cursor \(cursor)")
+                let response = try await fetchCursor(cursor)
+                
+                // if same cursor returned, loading is finished
+                if response.nextCursor == self.cursor {
+                    return .done(response.items)
+                }
+                
+                self.cursor = response.nextCursor
+                return .success(response.items)
+            } else {
+                page += 1
+                print("[\(Item.self) Fetcher] loading page \(page)")
+                let response = try await fetchPage(page)
+                
+                // if nothing returned, loading is finished
+                if response.items.isEmpty {
+                    return .done(.init())
+                }
```

Comment:
Could we assume that there are no more posts if `response.items < pageSize` rather than `response.items.isEmpty`? Doing so would make some pages load quicker - for a community with `< pageSize` posts, Mlem will currently wait for the second page load to be returned before displaying the first page of results. I don't think performing the second page load is necessary?

(This behavior also happened before this PR so feel free to leave OOS)

Reply:
That should work, since this check happens before any filtering--I've implemented and added some detailed logging just in case.